### DISCUSSION TITLE
Lesson 4 : Error while accessing array of struct

### DISCUSSION THREAD

**Discussion by CoggleChucker** [Cyfrin/foundry-full-course-cu#687] [Q&A] [answered]
While learning about Fundme contract, I tried to modify it using Struct
```
struct Funders
    {
        address wallet;
        uint256 amount;
    }
 ```
 created an array of struct at start of contract
 ```Funders[] public funders;```
 Populated it whenever new funder is added
```funders.push(Funders(msg.sender, msg.value));```

but in Withdraw function when i tried to reset, it throws error
```
function WithdrawFunds() public payable 
    {
        for(uint256 fundersIndex = 0; fundersIndex < funders.length; fundersIndex++)
        {           
            addressToAmount[funders[fundersIndex].wallet] = 0;
        }

        funders = new Funders[](0);

        (bool sendSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(sendSuccess,"Transfer failed");
    }
 ```
 Error Text:
 >UnimplementedFeatureError: Copying of type struct Fundme.Funders memory[] memory to storage not yet supported.

So, my question is:
**is there any way to access the array of custom(user-defined) structs in solidity?**

I tried accessing a single element. It works well
```
    function ShareFunder(uint256 _index) view public returns (address)
    {
        return(funders[_index].wallet);
    }
```

Here is my full code for fundme contract
```
//SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
contract Fundme
{
    struct Funders
    {
        address wallet;
        uint256 amount;
    }

    uint256 public minimumUSD = 5 * 1e18;
    Funders[] public funders;
    mapping (address => uint256) addressToAmount;

    function Fund() public payable
    {
        require(GetEthAmountInUSD(msg.value) > minimumUSD, "Not enough amount");
        funders.push(Funders(msg.sender, msg.value));
        addressToAmount[msg.sender] += msg.value;
    }

    function GetEthPrice() public view returns(uint256)
    {
        (,int256 price,,,) = AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306).latestRoundData();
        return uint256(price * 1e10);
    }

    function GetEthAmountInUSD(uint256 _amount)public view returns(uint256)
    {
        return ((GetEthPrice() * _amount) / 1e18);
    }

    function WithdrawFunds() public payable 
    {
        for(uint256 fundersIndex = 0; fundersIndex < funders.length; fundersIndex++)
        {           
            addressToAmount[funders[fundersIndex].wallet] = 0;
        }

        funders = new Funders[](0);

        (bool sendSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(sendSuccess,"Transfer failed");
    }

    function ShareFunder(uint256 _index) view public returns (address)
    {
        return(funders[_index].wallet);
    }

}
```

**Comment by theirrationalone** [comment]
Please remove `payable` modifier from `withdrawFunds` function first.. 

If you still getting error after this fix. Please share error too. 

Although you're accessing data from array correctly no doubt. This way you can access data from any array even of custom types. [Edited] [Edited] Sorry for losing context. 😅😅, But copying from memory to storage isn't supported yet.

**Reply by CoggleChucker** [reply]
The same error persists even after removing ```payable``` modifier.
>UnimplementedFeatureError: Copying of type struct Fundme.Funders memory[] memory to storage not yet supported.

**Reply by theirrationalone** [reply]
Heyy @CoggleChucker , Now i'm copying your code and testing it out and will try to find another way for your intention.

**Reply by CoggleChucker** [reply]
after doing some tweaking I found out the error is in the following line:
 ```
 funders = new Funders[](0);
 ```
 but I don't know how to solve it. (i.e. how to reset the array of struct inside a function)
 When I tried it with a simple address-type array it worked well. (as shown in the tutorial)

**Reply by theirrationalone** [reply]
Yes this is the problem, Finding another way, this isn't support in solidity yet.

**Reply by theirrationalone** [reply]
This is the only solution. I found... that fulfills your intensions....

```solidity
function WithdrawFunds() public {
        for (uint256 fundersIndex = 0; fundersIndex < funders.length; fundersIndex++) {
            addressToAmount[funders[fundersIndex].wallet] = 0;
            funders[fundersIndex].wallet = address(0);    // this one here <- ◀
            funders[fundersIndex].amount = 0;    // this one here <- ◀
        }

        (bool sendSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(sendSuccess, "Transfer failed");
    }
```

**Reply by CoggleChucker** [reply]
@theirrationalone Thank you for the solution. But this method has a big problem.
If I try to fund again after withdrawing, it will populate funders array from the last index. Rather than starting it from 0.

I am trying some solutions with ```array.pop()``` function.
will update you if I found anything.
Thank you

**Reply by theirrationalone** [reply, marked as answer]
My bad 😅, I forgot about array Indexing. As you're going to try `pop` method please keep in mind and verify if `pop` method also shifts the last element out from the array in solidity too. Or if `pop` method takes an index id to remove intended Element. 

If yes for last element deletion then, Please try to iterate through for loop in reverse (last to beginning) In this way you will never reach to index out of bound. 


E.G.: true if `pop` method deletes last element. 

```solidity
function WithdrawFunds() public {
    for (uint256 fundersIndex = funders.length -1; fundersIndex >= 0; fundersIndex--) {
        addressToAmount[funders[fundersIndex].wallet] = 0;
        funders.pop();
    }

(bool sendSuccess,) = payable(msg.sender).call{value: address(this).balance}("");

require(sendSuccess, "Transfer failed"); }
```

**Reply by CoggleChucker** [reply]
@theirrationalone Thank you.
just one change needed.
replacing ```uint256 fundersIndex``` with ```int256 fundersIndex```

otherwise ```fundersIndex``` will never go below 0

**Reply by theirrationalone** [reply]
@CoggleChucker ! You're So Welcome. Feeling productive to work with with, Woah, what the Great tiny but prominent information you introduced. Woah 🔥🔥🔥🔥🔥
